opentreemap/treemap/instance.py

```diff
@@ -768,9 +771,39 @@ def raise_errors(errors):
         if errors:
             raise_errors(errors)
 
-    def save(self, *args, **kwargs):
+    @transaction.atomic
+    def _bulk_assign_task_reference_numbers(self, tasks):
+        instance = Instance.objects.select_for_update().get(id=self.id)
+        for task in tasks:
+            task.reference_number = instance.task_sequence_number
+            instance.task_sequence_number += 1
+        instance.save(update_fields=['task_sequence_number'])
+
+    @transaction.atomic
+    def _bulk_assign_work_order_reference_numbers(self, work_orders):
+        instance = Instance.objects.select_for_update().get(id=self.id)
+        for work_order in work_orders:
+            work_order.reference_number = instance.work_order_sequence_number
+            instance.work_order_sequence_number += 1
+        instance.save(update_fields=['work_order_sequence_number'])
+
+    def save(self, tasks=None, work_orders=None, *args, **kwargs):
+        """
+        Keyword Arguments:
+
+        tasks - List of Task models
+        work_orders - List of WorkOrder models
+        """
         self.full_clean()
 
         self.url_name = self.url_name.lower()
 
+        if tasks:
```

Comment:
> Since there isn't any exception handling around this case, I'm wondering if it makes sense to get rid of the unique constraint?

I would say to keep the unique constraint - we can always show a "something went wrong, try again" toast to the user in that case.